projects/common/src/telemetry/user-telemetry-internal.service.ts

```diff
@@ -0,0 +1,155 @@
+import { Injectable, Injector, OnDestroy } from '@angular/core';
+import { NavigationEnd, Router } from '@angular/router';
+import { Observable, ReplaySubject, Subject } from 'rxjs';
+import { filter, map, takeUntil } from 'rxjs/operators';
+import { Dictionary } from '../utilities/types/types';
+import { UserTelemetryProvider, UserTelemetryRegistrationConfig, UserTraits } from './telemetry';
+
+@Injectable({ providedIn: 'root' })
+export class UserTelemetryInternalService implements OnDestroy {
+  private telemetryProviders: UserTelemetryInternalConfig[] = [];
+  private readonly destroyedSubject: ReplaySubject<void> = new ReplaySubject();
+
+  private readonly telemetryActionSubject: Subject<TelemetryAction> = new ReplaySubject();
+  private readonly telemetryAction$: Observable<TelemetryAction>;
+  private readonly identifyAction$: Observable<TelemetryAction>;
+  private readonly trackEventAction$: Observable<TelemetryAction>;
+  private readonly trackPageAction$: Observable<TelemetryAction>;
+  private readonly trackErrorAction$: Observable<TelemetryAction>;
+
+  public constructor(private readonly injector: Injector, private readonly router: Router) {
+    this.telemetryAction$ = this.telemetryActionSubject.pipe(takeUntil(this.destroyedSubject));
+
+    this.identifyAction$ = this.telemetryAction$.pipe(filter(action => action.type === TelemetryActionType.Identify));
+
+    this.trackEventAction$ = this.telemetryAction$.pipe(
+      filter(action => action.type === TelemetryActionType.TrackEvent)
+    );
+
+    this.trackPageAction$ = this.telemetryAction$.pipe(
+      filter(action => action.type === TelemetryActionType.TrackPageView)
+    );
+
+    this.trackErrorAction$ = this.telemetryAction$.pipe(
+      filter(action => action.type === TelemetryActionType.TrackError)
+    );
+
+    this.setupAutomaticPageTracking();
+    this.setupTelemetryActionHandlers();
+  }
+
+  public register(...configs: UserTelemetryRegistrationConfig<unknown>[]): void {
+    try {
+      const providers = configs.map(config => this.buildTelemetryProvider(config));
+      this.telemetryProviders = [...this.telemetryProviders, ...providers];
+    } catch (error) {
+      /**
+       * NoOp
+       */
+    }
+  }
+
+  public ngOnDestroy(): void {
+    this.destroyedSubject.next();
+    this.destroyedSubject.complete();
+  }
+
+  public identify(userTraits: UserTraits): void {
+    this.telemetryActionSubject.next({ type: TelemetryActionType.Identify, data: userTraits });
+  }
+
+  public shutdown(): void {
+    this.telemetryActionSubject.next({ type: TelemetryActionType.Shutdown });
+  }
+
+  public trackEvent(name: string, data: Dictionary<unknown>): void {
+    this.telemetryActionSubject.next({ type: TelemetryActionType.TrackEvent, name: name, data: data });
```

Comment:
Used observables to make event processing run in a separate execution loop.

Reply:
Is that something we actually need? We wouldn't do that when using them directly, as I assume they're not doing anything heavy.

Reply:
They make a network call and may add a small delay. I still want to protect against it.

Reply:
that network call would be out of band anyway though, right?